### PR TITLE
[PS-10] Feature/sponsorships removed at end of term

### PR DESCRIPTION
**Pull request by MGibson1**
## Type of change
- [ ] Bug fix
- [x] New feature development
- [ ] Tech debt (refactoring, code cleanup, dependency upgrades, etc)
- [ ] Build/deploy pipeline (DevOps)
- [ ] Other

## Objective
Jira ticket: https://bitwarden.atlassian.net/browse/PS-10

Goal is to not immediatly update stripe subscription on revoke/remove of a sponsorship. Instead, we want to **only** update sponsorships at sponsorship validation time. Remove/Revoke will just mark a sponsorship as `toDelete`, which is an invalid state that will trigger removal on validation (when invoice upcoming webhook is received).

## Code changes
* **Startup(s)**: I added globalSettings to the organization service collection extensions to choose which Revoke command to use. The only difference is when environment each type can delete sponsorships.
* **CancelSponsorshipCommand**: Base command is split into two: Delete and MarkToDelete. Delete is only called from the Validate command and Revoke in special cases.
* **CloudRevokeSponsorshipCommand**: Deletes if it hasn't been redeemed, otherwises marks for deletion
* **RemoveSponsorshipCommand**: Marks sponsorship for deletion. By definition these sponsorships are in use by a family or, so we mark to remove at next opportunity.
* **ValidateSponsorshipCommand**: Now houses the stripe subscription revert and mailing logics. Considers `ToDelete` sponsorships as invalid.
* **SelfHostedRevokeSponsorshipCommand**: Can delete only if the sponsorship hasn't yet been synced.

## Testing requirements
This will need testing along with UI implementations, but should not remove/revoke sponsorship in real time, but only once billing cycle ends.

## Before you submit
- [x] I have checked for formatting errors (`dotnet tool run dotnet-format --check`) (required)
- [ ] If making database changes - I have also updated Entity Framework queries and/or migrations
- [x] I have added **unit tests** where it makes sense to do so (encouraged but not required)
- [ ] This change requires a **documentation update** (notify the documentation team)
- [ ] This change has particular **deployment requirements** (notify the DevOps team)

### CODE REVIEW

**Review by jlf0dev — COMMENTED** [review]
Looks good to me, just a question

File: src/Core/OrganizationFeatures/OrganizationSponsorships/FamiliesForEnterprise/Cloud/CloudRevokeSponsorshipCommand.cs [hunk 1]
```diff
@@ -3,39 +3,32 @@
...
-        public async Task RevokeSponsorshipAsync(Organization sponsoredOrg, OrganizationSponsorship sponsorship)
+        public async Task RevokeSponsorshipAsync(OrganizationSponsorship sponsorship)
         {
             if (sponsorship == null)
             {
                 throw new BadRequestException("You are not currently sponsoring an organization.");
             }
 
-            if (sponsorship.SponsoredOrganizationId == null)
+            if (sponsorship.SponsoredOrganizationId == default)
```
Comment (justindbaur) [inline src/Core/OrganizationFeatures/OrganizationSponsorships/FamiliesForEnterprise/Cloud/CloudRevokeSponsorshipCommand.cs:24]: I prefer `null` here even though default would work just fine.

Comment (MGibson1) [reply src/Core/OrganizationFeatures/OrganizationSponsorships/FamiliesForEnterprise/Cloud/CloudRevokeSponsorshipCommand.cs:24]: oops misread this as `Sponsor**ing**` I agree.

File: src/Core/OrganizationFeatures/OrganizationSponsorships/FamiliesForEnterprise/CancelSponsorshipCommand.cs [hunk 2]
```diff
@@ -18,28 +18,25 @@ public class CancelSponsorshipCommand
             _organizationRepository = organizationRepository;
```
[the remark this answers is not in the document]
Comment (MGibson1) [reply src/Core/OrganizationFeatures/OrganizationSponsorships/FamiliesForEnterprise/CancelSponsorshipCommand.cs:18]: yup, done

**Review by justindbaur — APPROVED** [review]

**Pull request merged**